File: src/main/java/xyz/wagyourtail/jvmdg/ASMClassWriter.java
package xyz.wagyourtail.jvmdg;

import org.objectweb.asm.ClassWriter;
import xyz.wagyourtail.jvmdg.util.Function;

import java.util.ArrayList;
import java.util.List;

public class ASMClassWriter extends ClassWriter {
    private final Function<String, String> getSuperType;

    public ASMClassWriter(int flags, Function<String, String> getSuperType) {
        super(flags);
        this.getSuperType = getSuperType;
    }

    @Override
    protected String getCommonSuperClass(String type1, String type2) {
        List<String> l1 = getSuperTypes(type1);
        List<String> l2 = getSuperTypes(type2);
        // get intersection
        l1.retainAll(l2);
        // get first element
        return l1.get(0);
    }

    public List<String> getSuperTypes(String type) {
        List<String> l = new ArrayList<>();
        String current = type;
        while (!current.equals("java/lang/Object")) {
            l.add(current);
            try {
                Class<?> c = Class.forName(current.replace('/', '.'));
                current = c.getSuperclass().getName().replace('.', '/');
            } catch (ClassNotFoundException e) {
                current = getSuperType.apply(current);
                if (current == null) {
                    current = "java/lang/Object";
                    System.err.println("Could not find super type for " + type);
                }
            }
        }
        l.add("java/lang/Object");
        return l;
    }
}

File: gradle-plugin/src/main/kotlin/xyz/wagyourtail/jvmdg/gradle/ASMClassWriter.kt
package xyz.wagyourtail.jvmdg.gradle

import org.objectweb.asm.ClassWriter

class ASMClassWriter(flags: Int, private val getSuperType: (String) -> String?): ClassWriter(flags) {
    override fun getCommonSuperClass(type1: String, type2: String): String {
        val l1 = getSuperTypes(type1)
        val l2 = getSuperTypes(type2)
        return l1.intersect(l2.toSet()).first()
    }

    private fun getSuperTypes(type: String): MutableList<String> {
        val l: MutableList<String> = ArrayList()
        var current: String? = type
        while (current != "java/lang/Object") {
            l.add(current!!)
            try {
                val c = Class.forName(current.replace('/', '.'))
                current = c.superclass.getName().replace('.', '/')
            } catch (e: Throwable) {
                current = getSuperType(current)
                if (current == null) {
                    current = "java/lang/Object"
                    System.err.println("Could not find super type for $type")
                }
            }
        }
        l.add("java/lang/Object")
        return l
    }
}